File: Sources/structyC/04_linked_list/merge_lists.cpp
//
//  merge_lists.cpp
//  MyCLI
//
//  Created by new on 3/22/26.
//

class Node {
public:
    int val;
    Node* next;

    Node(int initialVal) {
        val = initialVal;
        next = nullptr;
    }
};

Node* mergeLists(Node* head1, Node* head2) {
    Node* dummy = new Node(-1);
    Node* tail = dummy;
    while (head1 != nullptr && head2 != nullptr) {
        if (head1->val <= head2->val) {
            tail->next = head1;
            head1 = head1->next;
        } else {
            tail->next = head2;
            head2 = head2->next;
        }
        tail = tail->next;
    }

    tail->next = head1 != nullptr ? head1 : head2;

    return dummy->next;
}

Node* mergeLists2(Node* head1, Node* head2) {
    if (head1 == nullptr) {
        return head2;
    }

    if (head2 != nullptr && head1->val > head2->val) {
        Node* tmp = head1;
        head1 = head2;
        head2 = tmp;
    }

    head1->next = mergeLists(head1->next, head2);

    return head1;
}

File: Sources/structy/04_linked_list/MergeLists.swift
//
//  MergeLists.swift
//  MyCLI
//
//  Created by new on 3/21/26.
//

func mergeLists2(_ head1: Node<Int>?, _ head2: Node<Int>?) -> Node<Int>? {
    guard let head1 else {
        return head2
    }

    guard let head2 else {
        return head1
    }

    if head1.val < head2.val {
        head1.next = mergeLists(head2, head1.next)
        return head1
    } else {
        head2.next = mergeLists(head1, head2.next)
        return head2
    }
}

func mergeLists(_ head1: Node<Int>?, _ head2: Node<Int>?) -> Node<Int>? {
    var tail = Node(0)
    let dummy = tail
    var p1 = head1
    var p2 = head2
    while let p1_ = p1, let p2_ = p2 {
        if p1_.val < p2_.val {
            tail.next = p1_
            p1 = p1_.next
        } else {
            tail.next = p2_
            p2 = p2_.next
        }
        tail = tail.next!
    }

    if p1 == nil {
        tail.next = p2
    } else {
        tail.next = p1
    }

    return dummy.next
}
